syntax = "proto3";
package vdss;

message VectorIdentifier { 
    oneof id {
        string uuid = 1;
        uint64 u64_id = 2;
    }
}

message Vector { 
    repeated float data = 1;
    uint32 dimension = 2;
}

message Payload {
    string json = 1;
}

message Status {
    int32 code = 1;
    string message = 2;
}

message HNSWConfig {
    uint32 m = 1;
    uint32 ef_construct = 2;
    uint32 ef_search = 3;
}

enum IndexDriver {
    FAISS = 0;
}

enum IndexAlgorithm {
    HNSW = 0;
}

enum StorageType {
    MEM = 0;
    DISK = 1;
}

enum DistanceMetric {
    COSINE = 0;
    L2 = 1;
}

message CollectionConfig {
    IndexDriver index_driver = 1;
    IndexAlgorithm index_algorithm = 2;
    StorageType storage_type = 3;
    uint32 dimension = 4;
    DistanceMetric distance_metric = 5;
    HNSWConfig hnsw_config = 6;
}

message CreateCollectionRequest {
    string collection_name = 1;
    CollectionConfig config = 2;
}

message UpsertVectorRequest {
    string collection_name = 1;
    VectorIdentifier vector_id = 2;
    Vector vector = 3;
}

message UpsertVectorResponse {
    Status status = 1;
}

message SearchRequest {
    string collection_name = 1;
    Vector vector = 2;
    int32 limit = 3;
}

message SearchResult {
    VectorIdentifier id = 1;
    float score = 2;
    Vector vector = 3;
    Payload payload = 4;
}

message SearchResponse {
    Status status = 1;
    repeated SearchResult results = 2;
}

message StatusResponse {
    int32 code = 1;
    string message = 2;
}

service VDSSService {
    rpc CreateCollection(CreateCollectionRequest) returns (StatusResponse);
    rpc UpsertVector(UpsertVectorRequest) returns (UpsertVectorResponse);
    rpc Search(SearchRequest) returns (SearchResponse);
}
